cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java

```diff
@@ -1011,14 +1018,18 @@ public void removeNullOrEmptyTags(final MetadataEntity metadataEntity) {
    * @return {@link MetadataChange} representing the metadata before the change and after
    */
   private MetadataChange setMetadata(MetadataEntry metadataEntry) {
-    Set<Indexer> indexersForKey = getIndexersForKey(metadataEntry.getKey());
+
     // get existing metadata
     Metadata existingMetadata = getMetadata(metadataEntry.getMetadataEntity());
-
-    // the entity does not have any metadata associated with it so we should index it with its type too
-    if (existingMetadata.getProperties().isEmpty() && existingMetadata.getTags().isEmpty()) {
-      indexersForKey.add(new MetadataEntityTypeIndexer());
+    Set<Indexer> indexersForKey;
+    // tags are always rewritten as a defined property key 'tag' so it is necessary that we always generate
```

Comment:
this comment confuses me...

Reply:
so after reading some code, I guess that what you are trying to do here is find out whether this entity existed before. And you do that by checking whether the properties of the existing metadata are empty. Now I have a couple questions:

- can it happen that an entity exists (and is indexed) but has no properties?
- what if I create an entity and then delete its properties and tags?

Anyway I think it would be a lot easier to understand the code if the Metadata class had a method found() or exists() that has the logic to determine whether the entity was found or not. Or perhaps the getMetadata should return null if it does not exist. It is pretty confusing and prone to break with future changes, if various places in the code make assumptions such as "for an existing entity there is always a property named 'tags'".

Reply:
Also does this mean, that ```Metadata.getProperties()``` always contains an entry for ```tags``` that has the same value as ```getTags()```? Shouldn't this implementation detail (that tags are stored as a property with a reserved name) be hidden from the client, that is, the ```tags``` property should not be in the property map.

Reply:
> Also does this mean, that Metadata.getProperties() always contains an entry for tags that has the same value as getTags()? Shouldn't this implementation detail (that tags are stored as a property with a reserved name) be hidden from the client, that is, the tags property should not be in the property map.

We hide this from clients. We never return the actual Map of key-value which we get from scan. We filter it out by removing the tags key. What we return to client is a Metadata object which differentiate between those.

Reply:
> this comment confuses me...
Agree. Even I had to wrap my head around to understand and make this work. Let me try to make it more cleaner.

Reply:
>can it happen that an entity exists (and is indexed) but has no properties?
what if I create an entity and then delete its properties and tags?
Anyway I think it would be a lot easier to understand the code if the Metadata class had a method found() or exists() that has the logic to determine whether the entity was found or not. Or perhaps the getMetadata should return null if it does not exist. It is pretty confusing and prone to break with future changes, if various places in the code make assumptions such as "for an existing entity there is always a property named 'tags'".

Question 1) It is possible if an entity is created and CDAP system did not generate any system metadata for it neither user tagged it with any metadata. In that case the entity is not discoverable by metadata search which is expected. 
Question 2) Deleting all properties and tags will delete all metadata Rows from the table so later when the entity itself is deleted or its parent is deleted there will be nothing to delete in the metadata system which sounds reasonable to me. Please correct me if I am missing something. 

The existence check will make the code a bit cleaner but has performance issue which adds up now since we emit metadata from programmatically and leads to drastic failures. Please see https://github.com/caskdata/cdap/pull/10385 in which we gained performance improvements by minimizing calls to HBase.

The retrieve call is needed to fetch existing metadata information for publishing audit trail. In current behavior the getMetadata only return empty properties and tags if no metadata was found for the entity i.e. there are no Rows in table and hence no Rows which needs to be deleted when the entity or its parent entity deletion happens.

Tag is a special case and causes a lot of confusion as we don't differentiate it from other key-value properties. We use the `isNewEntity` indexer every time when tag is written because it can be the case that `tag` is the only key-value pair which exists and the update of `tag` in it's current implementation deletes the `tag` row completely and then rewrites it as new record so this is like writing a new record every time tag is rewritten if there are no other properties then this is first record and hence needs the special indexing.

I have improved the comment on the doc to explain it better.